net.bobbo.player-controller/player_scripts/crouch.gd

```diff
@@ -33,4 +40,37 @@ func player_physics_process(delta: float):
 	var scale_difference = new_scale - previous_scale
 	
 	player.velocity.y += scale_difference * 0.5
-	player.height = new_scale
+	
+	# Make it so that the crouch can only go as low as regularly crouching and only high enough such that the player doesn't get stuck if an object is detected above them
+	player.height = clamp(new_scale, _max_player_height / 2, _height_above)
+	
+	# Get the 3D physics space from a Node3D like the player_pivot
+	var space_state = player.model_pivot.get_world_3d().direct_space_state
+	
+	# Calculate origin and ending of raycast
+	var origin = player.position
+	var end = player.position + Vector3(0, _max_player_height / 2, 0)
+	
+	# Create the raycast
+	var query = PhysicsRayQueryParameters3D.create(origin, end)
+	
+	# Exclude the player from raycast calculations and make the ray collide with areas
+	query.exclude = [player]
+	query.collide_with_areas = true
+	
+	# Cast the raycast
+	var result = space_state.intersect_ray(query)
+	
+	# If we detect an object at the top of the player
+	if result:
+		# Before this equation I subtracted by an offset of 0.001 so that way the height didn't exactly match so the player wouldn't get stuck, thankfully the equations I came up with seem to not need it
+		
+		# Calculate proper height to fit below object above the player
+		# Need to have a slightly different equation in case this happens to get proper crouch height
+		if(abs(player.position.y) > abs(result.position.y)):
```

Comment:
Not sure I follow this entirely... Why would there ever be a situation where the player's origin is higher than where the end of the raycast hits?

Reply:
Also I'm just realizing - we're relying on the Y axis always being up by doing this. We prob need some kind of way to check this relative to the up direction in PlayerController...

Reply:
Okay, so first of all since we're going with the boolean approach the equation stuff will become completely irrelevant since the entire point of it was to calculate the maximum player height when there is an obstruction stopping a player from achieving normal height. The reason why the player's origin would be higher than when the end of the raycasts ends is because there technically isn't notice how I take the absolute value of both the object above and the player's position. This is to ensure that we use the right equation if we're working with negative y coordinates since in that scenario with absolute value the player's y coordinate will be bigger than the object above them. I will admit I didn't consider the Y axis not being up, but in hindsight I should have anticipated that since now looking at it the Y axis will not always be up. I'll make the appropriate changes to account for the Y axis not being up but it likely won't include this equation because of the boolean approach.